Add, Subtract, Multiply Many

AyoobMH opened the pull request:
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature


* **What is the current behavior?** (You can also link to an open issue here)
 Addition, subtraction, and multiplication can only be done with single values.


* **What is the new behavior (if this is a feature change)?**
 Addition, subtraction, and multiplication using many values.


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
NO


* **Other information**:
 * Fixes #74 
 * Credit to @SepehrAsh for opening #78 ,but it got closed.

Rhymond commented:
Hi @AyoobMH 

Thanks for the PR! I was just thinking, maybe we can update original methods to accept variadic parameters instead?
```
func (m *Money) Add(om ...*Money) (*Money, error)
```

In that way the functionality to add multiple would exists and it would not introduce any breaking changes. Of course, I would expect to do a sanity check `if len(om) != 0` and return error before applying operation.

What do you think?

AyoobMH commented:
Hi @Rhymond, thanks for your efforts.
I think updating original methods to have variadic parameters is even better, it wouldn't introduce any redundant code. Are you thinking of updating only `Add`, `Subtract`, `Multiply`? I would be happy to open another PR for that.

Rhymond commented:
> Hi @Rhymond, thanks for your efforts.
> 
> I think updating original methods to have variadic parameters is even better, it wouldn't introduce any redundant code. Are you thinking of updating only `Add`, `Subtract`, `Multiply`? I would be happy to open another PR for that.

Exactly! Yes `Add`, `Subtract` and `Multiply`. Appreciate your time 🙏

AyoobMH commented:
Hi @Rhymond 

I have modified the PR and added variadic parameters. For checking parameters, I have used panic instead of an error, I believe not passing any parameter is considered a bad use of the package and should not happen during runtime. What do you think?

totemcaf commented on money.go:
```diff
@@ -198,26 +198,56 @@ func (m *Money) Negative() *Money {
-		return nil, err
+func (m *Money) Add(om ...*Money) (*Money, error) {
+	if len(om) == 0 {
+		panic("At least one Money is required to add")
```
I think you can return the original value instead of panic because it is not incorrect mathematically.

totemcaf commented on money.go:
```diff
@@ -198,26 +198,56 @@ func (m *Money) Negative() *Money {
-	return &Money{amount: mutate.calc.add(m.amount, om.amount), currency: m.currency}, nil
+	k := New(0, m.currency.Code)
+
+	for i := 0; i < len(om); i++ {
+		if err := m.assertSameCurrency(om[i]); err != nil {
+			return nil, err
+		}
+
+		k.amount = mutate.calc.add(k.amount, om[i].amount)
+	}
+
```
You can use the alternative form:
```suggestion
	for _, m2:= range om {
		if err := m.assertSameCurrency(m2); err != nil {
			return nil, err
		}

		k.amount = mutate.calc.add(k.amount, m2.amount)
	}

```

totemcaf commented on money.go:
```diff
@@ -198,26 +198,56 @@ func (m *Money) Negative() *Money {
-func (m *Money) Add(om *Money) (*Money, error) {
-	if err := m.assertSameCurrency(om); err != nil {
-		return nil, err
+func (m *Money) Add(om ...*Money) (*Money, error) {
```
It's typically to add an "s" to the name to denote an array of:
```suggestion
func (m *Money) Add(ms ...*Money) (*Money, error) {
```

totemcaf commented on money.go:
```diff
@@ -198,26 +198,56 @@ func (m *Money) Negative() *Money {
-		return nil, err
+func (m *Money) Subtract(om ...*Money) (*Money, error) {
+	if len(om) == 0 {
+		panic("At least one Money is required to subtract")
```
Same for add, you can return original value instead of panicking.

totemcaf commented on money.go:
```diff
@@ -198,26 +198,56 @@ func (m *Money) Negative() *Money {
-	return &Money{amount: mutate.calc.multiply(m.amount, mul), currency: m.currency}
+func (m *Money) Multiply(mul ...int64) *Money {
+	if len(mul) == 0 {
+		panic("At least one multiplier is required to multiply")
```
Here is more controversial to return the original value, so I think it is ok to panic

AyoobMH commented:
@totemcaf Thanks for the review, the requested changes were made!

AyoobMH commented:
Hi @Rhymond @totemcaf , this PR is ready for review, would appreciate your review, thanks in advance!

AyoobMH commented:
Hi @Rhymond, any updated for this PR, thanks in advance!

Rhymond reviewed (approved):
Sorry for late response, LGTM
